SpockOpaleye/src/TenantApi.hs

```diff
@@ -0,0 +1,133 @@
+create_tenant conn tenant@Tenant {tenant_id = id
+                                 ,tenant_name = name
+                                 ,tenant_firstname = first_name
+                                 ,tenant_lastname = last_name
+                                 ,tenant_email = email
+                                 ,tenant_phone = phone
+                                 ,tenant_status = status
+                                 ,tenant_ownerid = owner_id
+                                 ,tenant_backofficedomain = bo_domain} = do
+  ids <-
+    runInsertManyReturning
+      conn
+      tenantTable
+      (return
+         ( Nothing
+         , pgStrictText name
+         , pgStrictText first_name
+         , pgStrictText last_name
+         , pgStrictText email
+         , pgStrictText phone
+         , constant status
+         , toNullable . constant <$> owner_id
+         , pgStrictText bo_domain))
+      (\(id, _, _, _, _, _, _, _, _) -> id)
+  return $
+    case ids of
+      [] -> Nothing
+      (x:xs) ->
+        Just $
+        tenant
+        { tenant_id = x
+        }
+
+activate_tenant :: Connection -> TenantId -> IO GHC.Int.Int64
+activate_tenant conn tenant_id = set_tenant_status conn tenant_id TenantStatusActive
+
+deactivate_tenant :: Connection -> TenantId -> IO GHC.Int.Int64
+deactivate_tenant conn tenant_id = set_tenant_status conn tenant_id TenantStatusInActive
+
+set_tenant_status :: Connection -> TenantId -> TenantStatus -> IO GHC.Int.Int64
+set_tenant_status conn tenant_id status = update_tenant conn tenant_id update_func
+  where
+    update_func :: TenantTableR -> TenantTableW
+    update_func (id, name, fn, ln, em, ph, st, oid, bod) =
+      (Just id, name, fn, ln, em, ph, constant status, Just oid, bod)
+
+update_tenant :: Connection
+              -> TenantId
+              -> (TenantTableR -> TenantTableW)
+              -> IO GHC.Int.Int64
+update_tenant conn t_tenantid update_func =
+  runUpdate
+    conn
+    tenantTable
+    update_func
+    (\(id, _, _, _, _, _, _, _, _) -> id .== (constant t_tenantid))
+
+remove_tenant :: Connection -> Tenant -> IO GHC.Int.Int64
+remove_tenant conn Tenant {tenant_id = tid} = do
+  users_for_tenant <- read_users_for_tenant conn tid
+  roles_for_tenant <- read_roles_for_tenant conn tid
+  mapM_ (remove_role conn) roles_for_tenant
+  mapM_ (remove_user conn) users_for_tenant
+  runDelete
+    conn
+    tenantTable
+    (\(id, _, _, _, _, _, _, _, _) -> id .== (constant tid))
+
+read_tenants :: Connection -> IO [Tenant]
+read_tenants conn = do
+  r <- runQuery conn $ tenant_query
+  return $ fmap make_tenant r
+
+read_tenant_by_id :: Connection -> TenantId -> IO (Maybe Tenant)
+read_tenant_by_id conn id = do
+  r <- runQuery conn $ (tenant_query_by_id id)
+  return $
+    case r of
+      [] -> Nothing
+      rows -> Just $ Prelude.head $ fmap make_tenant rows
+
+make_tenant :: (TenantId, Text, Text, Text, Text, Text, TenantStatus, Maybe UserId, Text)
```

Comment:
should we expose this function to the outside world? Allow anyone to make a tenant with any status?